File: scripts/check_agent_setup.py
#!/usr/bin/env python3
"""Check the pinned skill snapshot and repo-local agent adapters, offline."""

from __future__ import annotations

import hashlib
import json
from pathlib import Path
import re
import sys


ROOT = Path(__file__).resolve().parents[1]


def check_setup(root: Path) -> list[str]:
    errors: list[str] = []
    lock_path = root / ".agents/matt-pocock-skills.lock.json"
    lock = json.loads(lock_path.read_text(encoding="utf-8"))
    if lock["version"] != 1:
        return ["Unsupported skill lock version"]
    if not re.fullmatch(r"[0-9a-f]{40}", lock["revision"]):
        errors.append("Lock revision must be a full commit SHA")
    if not lock["skills"]:
        errors.append("The vendored skill selection is empty")

    skills_root = root / ".agents/skills"
    names: set[str] = set()
    for skill in lock["skills"]:
        name = skill["name"]
        if not re.fullmatch(r"[a-z0-9]+(?:-[a-z0-9]+)*", name):
            errors.append(f"Invalid skill name: {name}")
            continue
        if name in names:
            errors.append(f"Duplicate skill in lock: {name}")
        names.add(name)
        folder = skills_root / name
        expected = skill["files"]
        if "SKILL.md" not in expected:
            errors.append(f"{name}: lock is missing SKILL.md")
        actual = {
            path.relative_to(folder).as_posix()
            for path in folder.rglob("*")
            if path.is_file() or path.is_symlink()
        }
        for missing in sorted(set(expected) - actual):
            errors.append(f"{name}: missing {missing}")
        for extra in sorted(actual - set(expected)):
            errors.append(f"{name}: unexpected file {extra}")
        for relative in sorted(actual & set(expected)):
            path = folder / relative
            if path.is_symlink():
                errors.append(f"{name}: vendored file is a symlink: {relative}")
            elif hashlib.sha256(path.read_bytes()).hexdigest() != expected[relative]:
                errors.append(f"{name}: modified {relative}")

    license_path = root / lock["license"]["path"]
    if not license_path.is_file():
        errors.append("Missing upstream license")
    elif hashlib.sha256(license_path.read_bytes()).hexdigest() != lock["license"]["sha256"]:
        errors.append("Upstream license differs from the recorded snapshot")

    canonical = {
        folder.name: folder
        for folder in skills_root.iterdir()
        if folder.is_dir()
    }
    for name, folder in sorted(canonical.items()):
        entrypoint = folder / "SKILL.md"
        if not entrypoint.is_file():
            errors.append(f"{name}: missing SKILL.md entrypoint")
            continue
        content = entrypoint.read_text(encoding="utf-8")
        frontmatter = re.match(r"\A---\r?\n(.*?)\r?\n---(?:\r?\n|\Z)", content, re.DOTALL)
        if not frontmatter:
            errors.append(f"{name}: missing YAML frontmatter")
            continue
        header = frontmatter.group(1)
        name_match = re.search(r"^name:[ \t]*(.+)$", header, re.MULTILINE)
        if not name_match or name_match.group(1).strip().strip("'\"") != name:
            errors.append(f"{name}: frontmatter name must match its directory")
        if not re.search(r"^description:[ \t]*\S", header, re.MULTILINE):
            errors.append(f"{name}: missing description")
        if re.search(r"^disable-model-invocation:[ \t]*true[ \t]*$", header, re.MULTILINE):
            policy = folder / "agents/openai.yaml"
            if not policy.is_file() or not re.search(
                r"^\s+allow_implicit_invocation:[ \t]*false[ \t]*$",
                policy.read_text(encoding="utf-8"), re.MULTILINE
            ):
                errors.append(f"{name}: missing matching explicit-only Codex policy")

        alias = root / ".claude/skills" / name
        if not alias.is_symlink() or alias.resolve() != folder.resolve():
            errors.append(f"{name}: Claude adapter must symlink to .agents/skills/{name}")

    adapters = root / ".claude/skills"
    if adapters.is_dir():
        for alias in adapters.iterdir():
            if alias.name not in canonical:
                errors.append(f"Claude adapter has no canonical skill: {alias.name}")

    if (root / "CLAUDE.md").read_text(encoding="utf-8").strip() != "@AGENTS.md":
        errors.append("CLAUDE.md must import the canonical AGENTS.md")
    for relative in (
        "AGENTS.md", "CONTEXT.md", "Project-Charter.md", "docs/architecture.md",
        "docs/engineering.md", "docs/agents/skills.md", "docs/agents/domain.md",
        "docs/agents/issue-tracker.md", "docs/agents/triage-labels.md",
    ):
        if not (root / relative).is_file():
            errors.append(f"Missing referenced guidance: {relative}")
    return errors


def main() -> int:
    try:
        errors = check_setup(ROOT)
    except (OSError, ValueError, KeyError, TypeError) as error:
        print(f"Agent setup check failed: {error}", file=sys.stderr)
        return 1
    if errors:
        for error in errors:
            print(f"ERROR: {error}", file=sys.stderr)
        return 1
    print("Agent setup OK: pinned skill files, license, metadata, and adapters match.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/check-agent-setup.ts
import { createHash } from 'node:crypto';
import {
  lstatSync,
  readFileSync,
  readdirSync,
  realpathSync,
  statSync,
} from 'node:fs';
import { join, resolve } from 'node:path';

function record(value: unknown, label: string): Record<string, unknown> {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new Error(label + ' must be an object');
  }
  return Object.fromEntries(Object.entries(value));
}
function text(value: unknown, label: string): string {
  if (typeof value !== 'string') throw new Error(label + ' must be a string');
  return value;
}
function isFile(path: string): boolean {
  return statSync(path, { throwIfNoEntry: false })?.isFile() ?? false;
}
function isDirectory(path: string): boolean {
  return statSync(path, { throwIfNoEntry: false })?.isDirectory() ?? false;
}
function isSymlink(path: string): boolean {
  return lstatSync(path, { throwIfNoEntry: false })?.isSymbolicLink() ?? false;
}
function digest(path: string): string {
  return createHash('sha256').update(readFileSync(path)).digest('hex');
}
function filesIn(folder: string, prefix = ''): string[] {
  if (!isDirectory(folder)) return [];
  const files: string[] = [];
  for (const entry of readdirSync(folder, { withFileTypes: true })) {
    const relative = prefix + entry.name;
    if (entry.isSymbolicLink() || entry.isFile()) files.push(relative);
    else if (entry.isDirectory())
      files.push(...filesIn(join(folder, entry.name), relative + '/'));
  }
  return files.sort();
}

/** Validate an installed skill snapshot and its repository adapters, offline. */
export function checkAgentSetup(root: string): string[] {
  const errors: string[] = [];
  const data: unknown = JSON.parse(
    readFileSync(join(root, '.agents/matt-pocock-skills.lock.json'), 'utf8'),
  );
  const lock = record(data, 'Skill lock');
  if (lock.version !== 1) return ['Unsupported skill lock version'];
  if (!/^[0-9a-f]{40}$/.test(text(lock.revision, 'Lock revision')))
    errors.push('Lock revision must be a full commit SHA');
  if (!Array.isArray(lock.skills))
    throw new Error('Lock skills must be an array');
  if (lock.skills.length === 0)
    errors.push('The vendored skill selection is empty');
  const skillsRoot = join(root, '.agents/skills');
  const names = new Set<string>();
  for (const item of lock.skills) {
    const skill = record(item, 'Locked skill');
    const name = text(skill.name, 'Skill name');
    if (!/^[a-z0-9]+(?:-[a-z0-9]+)*$/.test(name)) {
      errors.push('Invalid skill name: ' + name);
      continue;
    }
    if (names.has(name)) errors.push('Duplicate skill in lock: ' + name);
    names.add(name);
    const folder = join(skillsRoot, name);
    const expected = record(skill.files, name + ': locked files');
    if (!Object.hasOwn(expected, 'SKILL.md'))
      errors.push(name + ': lock is missing SKILL.md');
    const actual = new Set(filesIn(folder));
    for (const relative of Object.keys(expected).sort()) {
      if (!actual.has(relative)) errors.push(name + ': missing ' + relative);
    }
    for (const relative of actual) {
      const path = join(folder, relative);
      if (!Object.hasOwn(expected, relative))
        errors.push(name + ': unexpected file ' + relative);
      else if (isSymlink(path))
        errors.push(name + ': vendored file is a symlink: ' + relative);
      else if (digest(path) !== text(expected[relative], name + ': file hash'))
        errors.push(name + ': modified ' + relative);
    }
  }
  const license = record(lock.license, 'Locked license');
  const licensePath = join(root, text(license.path, 'License path'));
  if (!isFile(licensePath)) errors.push('Missing upstream license');
  else if (digest(licensePath) !== text(license.sha256, 'License hash'))
    errors.push('Upstream license differs from the recorded snapshot');
  const canonical = readdirSync(skillsRoot)
    .sort()
    .filter((name) => isDirectory(join(skillsRoot, name)));
  for (const name of canonical) {
    const folder = join(skillsRoot, name);
    const entrypoint = join(folder, 'SKILL.md');
    if (!isFile(entrypoint)) {
      errors.push(name + ': missing SKILL.md entrypoint');
      continue;
    }
    const content = readFileSync(entrypoint, 'utf8').replace(/\r\n?/g, '\n');
    const header = /^---\n([\s\S]*?)\n---(?:\n|$)/.exec(content)?.[1];
    if (header === undefined) {
      errors.push(name + ': missing YAML frontmatter');
      continue;
    }
    const declaredName = /^name:[ \t]*(.+)$/m
      .exec(header)?.[1]
      ?.trim()
      .replace(/^['"]+|['"]+$/g, '');
    if (declaredName !== name)
      errors.push(name + ': frontmatter name must match its directory');
    if (!/^description:[ \t]*\S/m.test(header))
      errors.push(name + ': missing description');
    if (/^disable-model-invocation:[ \t]*true[ \t]*$/m.test(header)) {
      const policy = join(folder, 'agents/openai.yaml');
      if (
        !isFile(policy) ||
        !/^\s+allow_implicit_invocation:[ \t]*false[ \t]*$/m.test(
          readFileSync(policy, 'utf8').replace(/\r\n?/g, '\n'),
        )
      ) {
        errors.push(name + ': missing matching explicit-only Codex policy');
      }
    }
    const alias = join(root, '.claude/skills', name);
    if (
      !isSymlink(alias) ||
      !isDirectory(alias) ||
      realpathSync(alias) !== realpathSync(folder)
    ) {
      errors.push(
        name + ': Claude adapter must symlink to .agents/skills/' + name,
      );
    }
  }
  const adapters = join(root, '.claude/skills');
  if (isDirectory(adapters)) {
    for (const name of readdirSync(adapters).sort()) {
      if (!canonical.includes(name))
        errors.push('Claude adapter has no canonical skill: ' + name);
    }
  }
  if (readFileSync(join(root, 'CLAUDE.md'), 'utf8').trim() !== '@AGENTS.md') {
    errors.push('CLAUDE.md must import the canonical AGENTS.md');
  }
  for (const relative of [
    'AGENTS.md',
    'CONTEXT.md',
    'Project-Charter.md',
    'docs/architecture.md',
    'docs/engineering.md',
    'docs/agents/skills.md',
    'docs/agents/domain.md',
    'docs/agents/issue-tracker.md',
    'docs/agents/triage-labels.md',
  ]) {
    if (!isFile(join(root, relative)))
      errors.push('Missing referenced guidance: ' + relative);
  }
  return errors;
}

if (import.meta.main) {
  try {
    const errors = checkAgentSetup(resolve(import.meta.dirname, '..'));
    if (errors.length) {
      for (const error of errors) console.error('ERROR: ' + error);
      process.exitCode = 1;
    } else
      console.log(
        'Agent setup OK: pinned skill files, license, metadata, and adapters match.',
      );
  } catch (error) {
    console.error(
      'Agent setup check failed: ' +
        (error instanceof Error ? error.message : String(error)),
    );
    process.exitCode = 1;
  }
}
